naga/src/back/msl/writer.rs

```diff
@@ -87,6 +104,62 @@ struct TypeContext<'a> {
     first_time: bool,
 }
 
+impl<'a> TypeContext<'a> {
+    fn zero_value_string(&self) -> String {
+        let ty = &self.gctx.types[self.handle];
+        match ty.inner {
+            crate::TypeInner::Scalar(scalar) | crate::TypeInner::Atomic(scalar) => {
+                scalar_zero_value_str(scalar).to_string()
+            }
+            crate::TypeInner::Vector { size, scalar } => {
+                format!(
+                    "{}::{}{}({})",
+                    NAMESPACE,
+                    scalar.to_msl_name(),
+                    back::vector_size_str(size),
+                    scalar_zero_value_str(scalar),
+                )
+            }
+            crate::TypeInner::Matrix { columns, rows, .. } => {
+                format!(
+                    "{}::{}{}x{}({})",
+                    NAMESPACE,
+                    crate::Scalar::F32.to_msl_name(),
+                    back::vector_size_str(columns),
+                    back::vector_size_str(rows),
+                    scalar_zero_value_str(crate::Scalar::F32),
+                )
+            }
+            crate::TypeInner::Pointer { .. } | crate::TypeInner::ValuePointer { .. } => {
+                "nullptr".to_string()
+            }
+            crate::TypeInner::Struct { .. } => "{{}}".to_string(),
+            _ => unimplemented!("Don't know how to produce a zero value for this type."),
+        }
+    }
+
+    fn scalar(&self) -> Option<crate::Scalar> {
+        let ty = &self.gctx.types[self.handle];
+        match ty.inner {
+            crate::TypeInner::Scalar(scalar) | crate::TypeInner::Atomic(scalar) => Some(scalar),
+            crate::TypeInner::Vector { scalar, .. } => Some(scalar),
+            crate::TypeInner::Matrix { .. } => Some(crate::Scalar::F32),
+            _ => None,
+        }
+    }
+
+    fn dimension(&self) -> u32 {
+        let ty = &self.gctx.types[self.handle];
+        match ty.inner {
+            crate::TypeInner::Vector { size, .. } => crate::back::vector_size_u32(size),
+            crate::TypeInner::Matrix { columns, rows, .. } => {
+                crate::back::vector_size_u32(columns) * crate::back::vector_size_u32(rows)
+            }
+            _ => 1,
+        }
+    }
```

Comment:
```suggestion
    fn vertex_input_dimension(&self) -> u32 {
        let ty = &self.gctx.types[self.handle];
        match ty.inner {
            crate::TypeInner::Scalar(_) => 1,
            crate::TypeInner::Vector { size, .. } => size as u32,
            _ => unreachable!(),
        }
    }
```

I think it's worth making this function more specialized/constrained since normally the type system is strict and we'd never treat scalars as vectors of size 1.